lib/pdfkit/source.rb

```diff
@@ -1,23 +1,45 @@
 class PDFKit
   class Source
+    SOURCE_FROM_STDIN = '-'
+
     def initialize(url_file_or_html)
       @source = url_file_or_html
     end
-    
+
     def url?
-      @source.is_a?(String) && @source.match(/\Ahttp/)
+      @is_url ||= @source.is_a?(String) && @source.match(/\Ahttp/)
     end
-    
+
     def file?
-      @source.kind_of?(File)
+      @is_file ||= @source.kind_of?(File)
     end
-    
+
     def html?
-      !(url? || file?)
+      @is_html ||= !(url? || file?)
     end
-    
+
+    def to_input_for_command
+      if file?
+        @source.path
+      elsif url?
+        %{"#{shell_safe_url}"}
+      else
+        SOURCE_FROM_STDIN
+      end
+    end
+
     def to_s
       file? ? @source.path : @source
     end
+
+    private
+
+    def shell_safe_url
+      url_needs_escaping? ? URI::escape(@source) : @source
+    end
+
+    def url_needs_escaping?
+      URI::decode(@source) == @source
```

Comment:
I'm not opposed to that, and I'm out of my element here, so apologies if I sound like a broken record repeating this question again:

I thought we determined that we wanted to support urls like:

```
https://www.google.com/search?q='cat<dev/zero>/dev/null'
```

Maybe I misread that in the discussion? It's what this PR uses as the example of something that needs escaping, but that we want to allow. I'm just not sure how to do that while not supporting:

```
"https://google.com/search?q=\"useradd -p secrete hacked\""
```

For example:

```
[14] pry(main)> url = "https://www.google.com/search?q='cat<dev/zero>/dev/null'"
=> "https://www.google.com/search?q='cat<dev/zero>/dev/null'"
[15] pry(main)> bad_url = "https://google.com/search?q=\"useradd -p secrete hacked\""
=> "https://google.com/search?q=\"useradd -p secrete hacked\""
[16] pry(main)> URI::parse(url)
URI::InvalidURIError: bad URI(is not URI?): https://www.google.com/search?q='cat<dev/zero>/dev/null'
from /Users/aunger/.rvm/rubies/ruby-2.0.0-p598/lib/ruby/2.0.0/uri/common.rb:176:in `split'
[17] pry(main)> URI::parse(bad_url)
URI::InvalidURIError: bad URI(is not URI?): https://google.com/search?q="useradd -p secrete hacked"
from /Users/aunger/.rvm/rubies/ruby-2.0.0-p598/lib/ruby/2.0.0/uri/common.rb:176:in `split'
[18] pry(main)> URI::parse(URI::escape(url))
=> #<URI::HTTPS:0x007ffbc1baf530 URL:https://www.google.com/search?q='cat%3Cdev/zero%3E/dev/null'>
[19] pry(main)> URI::parse(URI::escape(bad_url))
=> #<URI::HTTPS:0x007ffbc1b76fc8 URL:https://google.com/search?q=%22useradd%20-p%20secrete%20hacked%22>
```